<a href="https://colab.research.google.com/github/jieyingc/fast-failure-detection/blob/main/explore_dataset_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from datasets import load_dataset

dataset = load_dataset("SWE-bench/SWE-smith-trajectories")

print(dataset)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md: 0.00B [00:00, ?B/s]

data/tool-00000-of-00008.parquet:   0%|          | 0.00/249M [00:00<?, ?B/s]

data/tool-00001-of-00008.parquet:   0%|          | 0.00/116M [00:00<?, ?B/s]

data/tool-00002-of-00008.parquet:   0%|          | 0.00/99.6M [00:00<?, ?B/s]

data/tool-00003-of-00008.parquet:   0%|          | 0.00/138M [00:00<?, ?B/s]

data/tool-00004-of-00008.parquet:   0%|          | 0.00/77.8M [00:00<?, ?B/s]

data/tool-00005-of-00008.parquet:   0%|          | 0.00/123M [00:00<?, ?B/s]

data/tool-00006-of-00008.parquet:   0%|          | 0.00/153M [00:00<?, ?B/s]

data/tool-00007-of-00008.parquet:   0%|          | 0.00/122M [00:00<?, ?B/s]

data/xml-00000-of-00008.parquet:   0%|          | 0.00/106M [00:00<?, ?B/s]

data/xml-00001-of-00008.parquet:   0%|          | 0.00/121M [00:00<?, ?B/s]

data/xml-00002-of-00008.parquet:   0%|          | 0.00/115M [00:00<?, ?B/s]

data/xml-00003-of-00008.parquet:   0%|          | 0.00/144M [00:00<?, ?B/s]

data/xml-00004-of-00008.parquet:   0%|          | 0.00/126M [00:00<?, ?B/s]

data/xml-00005-of-00008.parquet:   0%|          | 0.00/244M [00:00<?, ?B/s]

data/xml-00006-of-00008.parquet:   0%|          | 0.00/113M [00:00<?, ?B/s]

data/xml-00007-of-00008.parquet:   0%|          | 0.00/107M [00:00<?, ?B/s]

data/ticks-00000-of-00008.parquet:   0%|          | 0.00/114M [00:00<?, ?B/s]

data/ticks-00001-of-00008.parquet:   0%|          | 0.00/94.6M [00:00<?, ?B/s]

data/ticks-00002-of-00008.parquet:   0%|          | 0.00/112M [00:00<?, ?B/s]

data/ticks-00003-of-00008.parquet:   0%|          | 0.00/143M [00:00<?, ?B/s]

data/ticks-00004-of-00008.parquet:   0%|          | 0.00/110M [00:00<?, ?B/s]

data/ticks-00005-of-00008.parquet:   0%|          | 0.00/95.7M [00:00<?, ?B/s]

data/ticks-00006-of-00008.parquet:   0%|          | 0.00/120M [00:00<?, ?B/s]

data/ticks-00007-of-00008.parquet:   0%|          | 0.00/229M [00:00<?, ?B/s]

Generating tool split:   0%|          | 0/24100 [00:00<?, ? examples/s]

Generating xml split:   0%|          | 0/26076 [00:00<?, ? examples/s]

Generating ticks split:   0%|          | 0/25826 [00:00<?, ? examples/s]

DatasetDict({
    tool: Dataset({
        features: ['messages', 'instance_id', 'resolved', 'model', 'traj_id', 'patch'],
        num_rows: 24100
    })
    xml: Dataset({
        features: ['messages', 'instance_id', 'resolved', 'model', 'traj_id', 'patch'],
        num_rows: 26076
    })
    ticks: Dataset({
        features: ['messages', 'instance_id', 'resolved', 'model', 'traj_id', 'patch'],
        num_rows: 25826
    })
})


In [3]:
sample = dataset["tool"][0]

for k, v in sample.items():
    print(k, type(k))
    ## print(v)

messages <class 'str'>
instance_id <class 'str'>
resolved <class 'str'>
model <class 'str'>
traj_id <class 'str'>
patch <class 'str'>


In [4]:
from collections import Counter

print(Counter(dataset["tool"]["resolved"]))
print(Counter(dataset["xml"]["resolved"]))
print(Counter(dataset["ticks"]["resolved"]))

Counter({False: 14673, True: 9427})
Counter({False: 14573, True: 11503})
Counter({False: 14396, True: 11430})


In [17]:
for split in dataset.keys():
    print(f"\n=== {split} ===")
    print(Counter(dataset[split]["model"]).most_common(10))


=== tool ===
[('claude-3-7-sonnet-20250219', 17715), ('claude-3-5-sonnet-20241022', 5751), ('gpt-4o-2024-08-06', 634)]

=== xml ===
[('claude-3-7-sonnet-20250219', 18457), ('claude-3-5-sonnet-20241022', 7091), ('gpt-4o-2024-08-06', 528)]

=== ticks ===
[('claude-3-7-sonnet-20250219', 18457), ('claude-3-5-sonnet-20241022', 6894), ('gpt-4o-2024-08-06', 475)]


In [18]:
tool_ids = set(dataset["tool"]["instance_id"])
xml_ids = set(dataset["xml"]["instance_id"])
ticks_ids = set(dataset["ticks"]["instance_id"])

print("tool ∩ xml:", len(tool_ids & xml_ids))
print("tool ∩ ticks:", len(tool_ids & ticks_ids))
print("xml ∩ ticks:", len(xml_ids & ticks_ids))

tool ∩ xml: 10637
tool ∩ ticks: 10636
xml ∩ ticks: 13500


- dataset 由 tool/xml/ticks 三个子集构成，不是标准 train/val/test
- 三个子集之间 instance_id 大量重叠，不能直接拿来当 train/val/test

### **Check null and empty**
- patch 字段有较高比例为空，但 empty patch 并不意味着 agent 没有完成修复
- resolved=True 的 empty-patch 样本仍然包含完整的修复与提交轨迹

In [19]:
for col in dataset["tool"].column_names:
    null_count = sum(row[col] is None for row in dataset["tool"])
    print(f"{col}: {null_count} nulls")

messages: 0 nulls
instance_id: 0 nulls
resolved: 0 nulls
model: 0 nulls
traj_id: 0 nulls
patch: 0 nulls


In [20]:
def check_empty_strings(hf_dataset):
    for col in hf_dataset.column_names:
        empty_count = 0
        for row in hf_dataset:
            value = row[col]
            if isinstance(value, str) and value.strip() == "":
                empty_count += 1
        print(f"{col}: {empty_count} empty strings")

print(check_empty_strings(dataset["tool"]))
print(check_empty_strings(dataset["xml"]))
print(check_empty_strings(dataset["ticks"]))

messages: 0 empty strings
instance_id: 0 empty strings
resolved: 0 empty strings
model: 0 empty strings
traj_id: 0 empty strings
patch: 5944 empty strings
None
messages: 0 empty strings
instance_id: 0 empty strings
resolved: 0 empty strings
model: 0 empty strings
traj_id: 0 empty strings
patch: 7352 empty strings
None
messages: 0 empty strings
instance_id: 0 empty strings
resolved: 0 empty strings
model: 0 empty strings
traj_id: 0 empty strings
patch: 7279 empty strings
None


In [21]:
empty_patch_labels = Counter()
nonempty_patch_labels = Counter()

for row in dataset["tool"]:
    if row["patch"].strip() == "":
        empty_patch_labels[row["resolved"]] += 1
    else:
        nonempty_patch_labels[row["resolved"]] += 1

print("Empty patch labels:", empty_patch_labels)
print("Non-empty patch labels:", nonempty_patch_labels)

Empty patch labels: Counter({False: 3662, True: 2282})
Non-empty patch labels: Counter({False: 11011, True: 7145})


In [32]:
import json

rows = [row for row in dataset["tool"] if row["patch"].strip() == "" and row["resolved"]]

for row in rows[:3]:
    msgs = json.loads(row["messages"])
    print("INSTANCE:", row["instance_id"])
    print("MODEL:", row["model"])
    print("NUM_MESSAGES:", len(msgs))
    print("LAST 3 MESSAGES:")
    for m in msgs[-3:]:
       print(m)
    print("=" * 80)

INSTANCE: django-money__django-money.835c1ab8.func_pm_ctrl_shuffle__viqnyl9u
MODEL: claude-3-7-sonnet-20250219
NUM_MESSAGES: 49
LAST 3 MESSAGES:
{'role': 'assistant', 'content': 'Let me remove the reproduction script and submit again:', 'thought': 'Let me remove the reproduction script and submit again:', 'action': 'rm /testbed/reproduce_issue.py', 'agent': 'main', 'tool_calls': [{'index': 1, 'function': {'arguments': '{"command": "rm /testbed/reproduce_issue.py"}', 'name': 'bash'}, 'id': 'toolu_01RX7JN3tbDwDNYoAFzeEDaL', 'type': 'function'}], 'message_type': 'action'}
{'role': 'tool', 'content': [{'type': 'text', 'text': 'Your command ran successfully and did not produce any output.'}], 'agent': 'main', 'message_type': 'observation', 'tool_call_ids': ['toolu_01RX7JN3tbDwDNYoAFzeEDaL'], 'cache_control': {'type': 'ephemeral'}}
{'role': 'assistant', 'content': '', 'thought': '', 'action': 'submit', 'agent': 'main', 'tool_calls': [{'index': 0, 'function': {'arguments': '{}', 'name': 'subm

**split by instance_id not by row**:
- tool 子集中有 10,637 个 unique instances 和 24,100 条 trajectories，说明一个 instance 可对应多次 agent 尝试

In [23]:
tool_instance_counts = Counter(dataset["tool"]["instance_id"])
dup_tool = sum(c > 1 for c in tool_instance_counts.values())
max_count = max(tool_instance_counts.values())

print("unique instance_ids:", len(tool_instance_counts))
print("duplicated instance_ids:", dup_tool)
print("max trajectories per instance:", max_count)

unique instance_ids: 10637
duplicated instance_ids: 4675
max trajectories per instance: 21


**Feature extraction**

In [18]:
import json
import re


def extract_features(row):
    patch = row["patch"] if row["patch"] is not None else ""
    patch_stripped = patch.strip()

    try:
        msgs = json.loads(row["messages"])
    except Exception:
        msgs = []

    # -------------------------
    # 1) Patch-level features
    # -------------------------
    patch_is_empty = int(patch_stripped == "")
    patch_char_len = len(patch)

    # Count added / removed lines (excluding diff headers like +++ / ---)
    num_lines_added = 0
    num_lines_removed = 0

    for line in patch.splitlines():
        if line.startswith("+++ ") or line.startswith("--- "):
            continue
        if line.startswith("+"):
            num_lines_added += 1
        elif line.startswith("-"):
            num_lines_removed += 1

    # Number of files touched
    num_files_touched = patch.count("diff --git")

    # Whether test files were modified
    touched_files = re.findall(r"diff --git a/(.*?) b/(.*?)\n", patch)
    test_file_modified = 0
    for a_path, b_path in touched_files:
        path = b_path.lower()
        if (
            "test" in path
            or "tests/" in path
            or path.endswith("_test.py")
            or "/test_" in path
        ):
            test_file_modified = 1
            break

    # Repeated-block (if the same added/removed code line repeats 3 times)
    code_lines = []
    for line in patch.splitlines():
        if line.startswith("+++ ") or line.startswith("--- ") or line.startswith("@@"):
            continue
        if line.startswith("+") or line.startswith("-"):
            content = line[1:].strip()
            if content:
                code_lines.append(content)

    contain_repeated_blocks = 0
    if code_lines:
        line_freq = {}
        for x in code_lines:
            line_freq[x] = line_freq.get(x, 0) + 1
        if max(line_freq.values()) >= 3:
            contain_repeated_blocks = 1

    # -------------------------
    # 2) Message / trajectory features
    # -------------------------
    num_messages = len(msgs)
    messages_char_len = len(row["messages"])

    num_assistant_messages = 0
    num_user_messages = 0
    num_tool_messages = 0
    num_system_messages = 0

    num_action_messages = 0
    num_observation_messages = 0

    has_submit = 0
    num_submit_calls = 0

    has_bash = 0
    num_bash_calls = 0

    has_str_replace_editor = 0

    num_edit_calls = 0
    num_error_messages = 0

    edit_tool_names = {"str_replace_editor", "edit_file", "create_file", "insert", "replace", "write_file"}

    error_keywords = ["error", "exception", "traceback", "failed", "failure", "syntaxerror", "typeerror", "valueerror", "assertionerror"]

    for msg in msgs:
        # Role counts
        role = msg.get("role", "")

        if role == "assistant":
            num_assistant_messages += 1
        elif role == "user":
            num_user_messages += 1
        elif role == "tool":
            num_tool_messages += 1
        elif role == "system":
            num_system_messages += 1

        # Message-type counts
        message_type = msg.get("message_type", "")

        if message_type == "action":
            num_action_messages += 1
        elif message_type == "observation":
            num_observation_messages += 1

        # Check action field
        action_text = str(msg.get("action", "")).lower()

        if "submit" in action_text:
            has_submit = 1

        if "bash" in action_text:
            has_bash = 1

        if "str_replace_editor" in action_text:
            has_str_replace_editor = 1

        # tool_calls parsing

        tool_calls = msg.get("tool_calls", [])
        if isinstance(tool_calls, list):
            for tc in tool_calls:
                fn_name = (
                    tc.get("function", {}).get("name", "")
                    if isinstance(tc, dict)
                    else ""
                )
                fn_name = str(fn_name).lower()

                if fn_name == "submit":
                    has_submit = 1
                    num_submit_calls += 1

                if fn_name == "bash":
                    has_bash = 1
                    num_bash_calls += 1

                if fn_name == "str_replace_editor":
                    has_str_replace_editor = 1

                if fn_name in edit_tool_names or fn_name == "str_replace_editor":
                    num_edit_calls += 1

        # error keyword counting

        text_blobs = [] # Collect all text (content, thought, action)

        content = msg.get("content", "")
        if isinstance(content, str):
            text_blobs.append(content.lower())
        elif isinstance(content, list):
            for item in content:
                if isinstance(item, dict):
                    t = item.get("text", "")
                    if isinstance(t, str):
                        text_blobs.append(t.lower())

        thought = msg.get("thought", "")
        if isinstance(thought, str):
            text_blobs.append(thought.lower())

        if isinstance(action_text, str):
            text_blobs.append(action_text)

        joined_text = " ".join(text_blobs)
        for kw in error_keywords:
            num_error_messages += joined_text.count(kw)


    features = {
        # IDs / metadata
        "instance_id": row["instance_id"],
        "traj_id": row["traj_id"],
        "model": row["model"],

        # Label
        "resolved": int(bool(row["resolved"])),

        # Patch features
        "patch_is_empty": patch_is_empty,
        "patch_char_len": patch_char_len,
        "num_lines_added": num_lines_added,
        "num_lines_removed": num_lines_removed,
        "num_files_touched": num_files_touched,
        "test_file_modified": test_file_modified,
        "contain_repeated_blocks": contain_repeated_blocks,

        # Message features
        "num_messages": num_messages,
        "messages_char_len": messages_char_len,
        "assistant_num_messages": num_assistant_messages,
        "user_num_messages": num_user_messages,
        "tool_num_messages": num_tool_messages,
        "system_num_messages": num_system_messages,
        "num_action_messages": num_action_messages,
        "num_observation_messages": num_observation_messages,
        "has_submit": has_submit,
        "num_submit_calls": num_submit_calls,
        "has_bash": has_bash,
        "num_bash_calls": num_bash_calls,
        "has_str_replace_editor": has_str_replace_editor,
        "num_edit_calls": num_edit_calls,
        "num_error_messages": num_error_messages,
    }

    return features

In [19]:
import pandas as pd
from tqdm import tqdm

feature_rows = []

for i in tqdm(range(len(dataset["tool"]))):
    row = dataset["tool"][i]
    feature_rows.append(extract_features(row))

df = pd.DataFrame(feature_rows)

print(df.shape)
df.head()

100%|██████████| 24100/24100 [02:05<00:00, 191.99it/s]


(24100, 26)


,instance_id,traj_id,model,resolved,patch_is_empty,patch_char_len,num_lines_added,num_lines_removed,num_files_touched,test_file_modified,...,system_num_messages,num_action_messages,num_observation_messages,has_submit,num_submit_calls,has_bash,num_bash_calls,has_str_replace_editor,num_edit_calls,num_error_messages
0,django-money__django-money.835c1ab8.func_pm_ct...,django-money__django-money.835c1ab8.func_pm_ct...,claude-3-7-sonnet-20250219,1,1,0,0,0,0,0,...,1,24,24,1,2,1,10,1,12,61
1,marshmallow-code__apispec.8b421526.func_pm_rem...,marshmallow-code__apispec.8b421526.func_pm_rem...,claude-3-7-sonnet-20250219,1,0,1925,15,12,1,0,...,1,21,21,1,2,1,10,1,9,22
2,mozillazg__python-pinyin.e42dede5.func_pm_remo...,mozillazg__python-pinyin.e42dede5.func_pm_remo...,claude-3-7-sonnet-20250219,1,1,0,0,0,0,0,...,1,32,32,1,2,1,17,1,13,63
3,marshmallow-code__marshmallow.9716fc62.func_pm...,marshmallow-code__marshmallow.9716fc62.func_pm...,claude-3-7-sonnet-20250219,1,0,14794,4,261,3,0,...,1,24,24,1,2,1,11,1,11,132
4,luozhouyang__python-string-similarity.115acaac...,luozhouyang__python-string-similarity.115acaac...,claude-3-7-sonnet-20250219,1,1,0,0,0,0,0,...,1,24,24,1,2,1,12,1,10,30


In [20]:
print(df["resolved"].value_counts())
print(df["patch_is_empty"].value_counts())
print(df.isnull().sum().sort_values(ascending=False).head(10))

resolved
0    14673
1     9427
Name: count, dtype: int64
patch_is_empty
0    18156
1     5944
Name: count, dtype: int64
instance_id           0
traj_id               0
model                 0
resolved              0
patch_is_empty        0
patch_char_len        0
num_lines_added       0
num_lines_removed     0
num_files_touched     0
test_file_modified    0
dtype: int64


In [21]:
df.describe()

,resolved,patch_is_empty,patch_char_len,num_lines_added,num_lines_removed,num_files_touched,test_file_modified,contain_repeated_blocks,num_messages,messages_char_len,...,system_num_messages,num_action_messages,num_observation_messages,has_submit,num_submit_calls,has_bash,num_bash_calls,has_str_replace_editor,num_edit_calls,num_error_messages
count,24100.000000,24100.000000,2.410000e+04,2.410000e+04,24100.000000,24100.000000,24100.000000,24100.000000,24100.000000,24100.000000,...,24100.0,24100.000000,24100.000000,24100.000000,24100.00000,24100.000000,24100.000000,24100.000000,24100.000000,24100.000000
mean,0.391162,0.246639,3.252244e+04,6.965167e+02,57.321452,4.259461,0.083817,0.249129,64.170954,122135.757510,...,1.0,11.550124,11.550124,0.345726,0.59473,0.377303,4.441909,0.378257,6.478714,130.855851
std,0.488021,0.431064,2.606425e+06,5.607106e+04,4593.151761,207.849850,0.277120,0.432517,38.008198,83018.091752,...,0.0,18.990446,18.990446,0.475614,0.97682,0.484722,7.852592,0.484962,11.525265,203.277446
min,0.000000,0.000000,0.000000e+00,0.000000e+00,0.000000,0.000000,0.000000,0.000000,7.000000,7951.000000,...,1.0,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,3.000000
25%,0.000000,0.000000,3.780000e+02,1.000000e+00,0.000000,1.000000,0.000000,0.000000,37.000000,61789.750000,...,1.0,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,37.000000
50%,0.000000,0.000000,1.405000e+03,1.100000e+01,1.000000,1.000000,0.000000,0.000000,53.000000,98799.500000,...,1.0,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,77.000000
75%,1.000000,0.000000,3.003000e+03,3.700000e+01,5.000000,2.000000,0.000000,0.000000,83.000000,158348.250000,...,1.0,19.000000,19.000000,1.000000,1.00000,1.000000,7.000000,1.000000,10.000000,156.000000
max,1.000000,1.000000,3.928155e+08,8.340617e+06,708045.000000,30611.000000,1.000000,1.000000,303.000000,663706.000000,...,1.0,151.000000,151.000000,1.000000,64.00000,1.000000,144.000000,1.000000,117.000000,8378.000000


In [22]:
from sklearn.model_selection import GroupShuffleSplit

# train (80%) / val (10%) / test (10%)
gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
train_idx, temp_idx = next(gss.split(df, groups=df["instance_id"]))

train_df = df.iloc[train_idx].reset_index(drop=True)
temp_df = df.iloc[temp_idx].reset_index(drop=True)


gss2 = GroupShuffleSplit(n_splits=1, test_size=0.5, random_state=42)
val_idx, test_idx = next(gss2.split(temp_df, groups=temp_df["instance_id"]))

val_df = temp_df.iloc[val_idx].reset_index(drop=True)
test_df = temp_df.iloc[test_idx].reset_index(drop=True)

print("train shape:", train_df.shape)
print("val shape:", val_df.shape)
print("test shape:", test_df.shape)

train shape: (19321, 26)
val shape: (2434, 26)
test shape: (2345, 26)


In [23]:
train_instances = set(train_df["instance_id"])
val_instances = set(val_df["instance_id"])
test_instances = set(test_df["instance_id"])

print("train ∩ val:", len(train_instances & val_instances))
print("train ∩ test:", len(train_instances & test_instances))
print("val ∩ test:", len(val_instances & test_instances))

train ∩ val: 0
train ∩ test: 0
val ∩ test: 0


In [24]:
import numpy as np

log_features = [
    "patch_char_len",
    "num_lines_added",
    "num_lines_removed",
    "num_files_touched",
    "num_messages",
    "messages_char_len",
    "num_submit_calls",
    "num_bash_calls",
    "num_edit_calls",
    "num_error_messages",
]

for col in log_features:
    train_df[col] = np.log1p(train_df[col])
    val_df[col] = np.log1p(val_df[col])
    test_df[col] = np.log1p(test_df[col])

In [25]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

feature_cols = [
    "patch_is_empty",
    "patch_char_len",
    "num_lines_added",
    "num_lines_removed",
    "num_files_touched",
    "test_file_modified",
    "contain_repeated_blocks",
    "num_messages",
    "messages_char_len",
    "assistant_num_messages",
    "user_num_messages",
    "tool_num_messages",
    "system_num_messages",
    "num_action_messages",
    "num_observation_messages",
    "has_submit",
    "num_submit_calls",
    "has_bash",
    "num_bash_calls",
    "has_str_replace_editor",
    "num_edit_calls",
    "num_error_messages",
]

X_train = train_df[feature_cols]
y_train = train_df["resolved"]

X_val = val_df[feature_cols]
y_val = val_df["resolved"]

X_test = test_df[feature_cols]
y_test = test_df["resolved"]

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)
X_test_scaled = scaler.transform(X_test)

clf = LogisticRegression(
    max_iter=2000,
    class_weight="balanced",
    random_state=42
)

clf.fit(X_train_scaled, y_train)

LogisticRegression(class_weight='balanced', max_iter=2000, random_state=42)

In [26]:
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, classification_report

val_pred = clf.predict(X_val_scaled)
val_prob = clf.predict_proba(X_val_scaled)[:, 1]

print("Validation Accuracy:", accuracy_score(y_val, val_pred))
print("Validation F1:", f1_score(y_val, val_pred))
print("Validation ROC-AUC:", roc_auc_score(y_val, val_prob))
print(classification_report(y_val, val_pred))

Validation Accuracy: 0.6224322103533279
Validation F1: 0.583219954648526
Validation ROC-AUC: 0.6701011985526911
              precision    recall  f1-score   support

           0       0.73      0.59      0.65      1474
           1       0.52      0.67      0.58       960

    accuracy                           0.62      2434
   macro avg       0.62      0.63      0.62      2434
weighted avg       0.65      0.62      0.63      2434



In [27]:
test_pred = clf.predict(X_test_scaled)
test_prob = clf.predict_proba(X_test_scaled)[:, 1]

print("Test Accuracy:", accuracy_score(y_test, test_pred))
print("Test F1:", f1_score(y_test, test_pred))
print("Test ROC-AUC:", roc_auc_score(y_test, test_prob))
print(classification_report(y_test, test_pred))

Test Accuracy: 0.5808102345415779
Test F1: 0.5131253095591877
Test ROC-AUC: 0.6401645235354186
              precision    recall  f1-score   support

           0       0.75      0.54      0.63      1553
           1       0.42      0.65      0.51       792

    accuracy                           0.58      2345
   macro avg       0.59      0.60      0.57      2345
weighted avg       0.64      0.58      0.59      2345



In [28]:
coef_df = pd.DataFrame({
    "feature": feature_cols,
    "coefficient": clf.coef_[0]
}).sort_values("coefficient", ascending=False)

coef_df

,feature,coefficient
18,num_bash_calls,0.554463
16,num_submit_calls,0.209328
21,num_error_messages,0.123297
15,has_submit,0.083425
7,num_messages,0.078240
1,patch_char_len,0.054320
6,contain_repeated_blocks,0.053983
5,test_file_modified,0.044589
11,tool_num_messages,0.010053
13,num_action_messages,0.003483
